planet/subscription_request.py

```diff
@@ -0,0 +1,447 @@
+    ...                     [139.5648193359375,35.42374884923695]]],
+    ...     "type": "Polygon"
+    ... }
+    >>> source = catalog_source(
+    ...     ["PSScene"], ["ortho_analytic_4b"], geom, datetime(2021,3,1))
+
+    >>> delivery = amazon_s3(
+    ...     ACCESS_KEY_ID, SECRET_ACCESS_KEY, "test", "us-east-1")
+    ...
+    >>> subscription_request = build_request(
+    ...     'test_subscription', source, delivery)
+    ...
+
+    ```
+
+    Parameters:
+        name: Name of the subscription.
+        source: A source for the subscription, i.e. catalog.
+        delivery: A delivery mechanism e.g. GCS, AWS, Azure, or OCS.
+        notifications: Specify notifications via email/webhook
+        tools: Tools to apply to the products. Order defines
+            the toolchain order of operatations.
+    """
+    details = {"name": name, "source": source, "delivery": delivery}
+
+    if notifications:
+        details['notifications'] = notifications
+
+    if tools:
+        details['tools'] = tools
+
+    return details
+
+
+def catalog_source(
+    item_types: List[str],
+    asset_types: List[str],
+    geometry: dict,
+    start_time: datetime,
+    filter: Optional[dict] = None,
+    end_time: Optional[datetime] = None,
+    rrule: Optional[str] = None,
+) -> dict:
+    '''Catalog subscription source.
+
+    Parameters:
+    item_types: The class of spacecraft and processing level of the
+        subscription's matching items, e.g. PSScene.
+    asset_types: The data products which will be delivered for all subscription
+        matching items. An item will only match and deliver if all specified
+        asset types are published for that item.
+    geometry: The area of interest of the subscription that will be used to
+        determine matches.
+    start_time: The start time of the subscription. This time can be in the
+        past or future.
+    filter: The filter criteria based on item-level metadata.
+    end_time: The end time of the subscription. This time can be in the past or
+        future, and must be after the start_time.
+    rrule: The recurrence rule, given in iCalendar RFC 5545 format. Only
+        monthly recurrences are supported at this time.
+
+    Raises:
+        planet.exceptions.ClientError: If start_time or end_time are not valid
+            datetimes
+    '''
+    parameters = {
+        "item_types": item_types,
+        "asset_types": asset_types,
+        "geometry": geometry,
+    }
+
+    try:
+        parameters['start_time'] = _datetime_to_rfc3339(start_time)
+    except AttributeError:
+        raise ClientError('Could not convert start_time to an iso string')
+
+    if filter:
+        parameters['filter'] = filter
+
+    if end_time:
+        try:
+            parameters['end_time'] = _datetime_to_rfc3339(end_time)
+        except AttributeError:
+            raise ClientError('Could not convert end_time to an iso string')
+
```

Comment:
Why did you choose to not let if break it end_time is before start_time?

For instance, this command works:
```
> subscription_request.catalog_source("PSScene", "visual", {"a":5}, start_time=datetime.datetime(2023, 2, 13), end_time=datetime.datetime(2022, 2, 13))
> {'type': 'catalog',
 'parameters': {'item_types': 'PSScene',
  'asset_types': 'visual',
  'geometry': {'a': 5},
  'start_time': '2023-02-13T00:00:00Z',
  'end_time': '2022-02-13T00:00:00Z'}}
```

Reply:
We could absolutely go into depth for validating all parameters. It's a balance. And end time before start time is one of those places. I stopped at validating the format just because the chances of the user messing up the format are high, so it's nice to filter those out before hitting the API, while the chances of the user swapping start time and end time are low (more intuitive) so I chose let the API take the hit and send them the error.